### PR TITLE
Clean up CodepointWidthDetector

### PR DESCRIPTION
**Pull request by lhecker**
My long-term plan is to replace the `CodepointWidth` enum with a simple integer
return value that indicates the amount of columns a codepoint is wide.
This is necessary so that we can return 0 for ZWJs (zero width joiners).

This initial commit represents a cleanup effort around `CodepointWidthDetector`.
Since less code runs faster, this change has the nice side-effect of running
roughly 5-10% faster across the board. It also drops the binary size by ~1.2kB.

## Validation Steps Performed
* `CodepointWidthDetectorTests` passes ✅
* U+26bf (``"`u{26bf}"`` inside pwsh) is a wide glyph
  in OpenConsole and narrow one in Windows Terminal ✅

### CODE REVIEW

**Comment by DHowett** [comment]
> simple integer return value that indicates the amount of columns a codepoint is wide.

Accounting for ambiguity will be difficult!

File: src/types/GlyphWidth.cpp [hunk 1]
```diff
@@ -35,9 +35,9 @@ bool IsGlyphFullWidth(const wchar_t wch) noexcept
 // - pfnFallback - the function to use as the fallback method.
 // Return Value:
 // - <none>
-void SetGlyphWidthFallback(std::function<bool(const std::wstring_view)> pfnFallback)
+void SetGlyphWidthFallback(std::function<bool(const std::wstring_view&)> pfnFallback) noexcept
```
Comment (DHowett) [inline src/types/GlyphWidth.cpp:38]: i thought taking string views by reference was more expensive than by copy (because of the ABI or whatever)

File: src/host/ut_host/CodepointWidthDetectorTests.cpp [hunk 2]
```diff
@@ -84,15 +70,15 @@ class CodepointWidthDetectorTests
         VERIFY_ARE_EQUAL(0u, widthDetector._fallbackCache.size());
 
         // Lookup ambiguous width character.
-        widthDetector.IsWide(ambiguous);
+        widthDetector.IsWide(L"\x414");
```
Comment (DHowett) [inline src/host/ut_host/CodepointWidthDetectorTests.cpp:73]: huh?

File: src/host/srvinit.cpp [hunk 3]
```diff
@@ -5,30 +5,22 @@
...
-
-#include "../types/inc/GlyphWidth.hpp"
-
-#include "../server/DeviceHandle.h"
-#include "../server/Entrypoints.h"
-#include "../server/IoSorter.h"
-
-#include "../interactivity/inc/ISystemConfigurationProvider.hpp"
-#include "../interactivity/inc/ServiceLocator.hpp"
+#include "../inc/conint.h"
```
Comment (DHowett) [inline src/host/srvinit.cpp:14]: this sort seems unrelated

Comment (DHowett) [reply src/host/srvinit.cpp:14]: actually the changes to this file seem unrelated

Comment (lhecker) [reply src/host/srvinit.cpp:14]: Ah woops, I originally changed the signature of `SetGlyphWidthFallback`, but then reverted it to make the PR smaller.

File: src/cascadia/PublicTerminalCore/HwndTerminal.cpp [hunk 4]
```diff
@@ -24,25 +24,6 @@ static constexpr bool _IsMouseMessage(UINT uMsg)
...
-{
-    // use C++11 magic statics to make sure we only do this once.
-    static auto initialized = []() {
-        // *** THIS IS A SINGLETON ***
-        SetGlyphWidthFallback(_IsGlyphWideForceNarrowFallback);
-
-        return true;
-    }();
-    return initialized;
-}
```
Comment (lhecker) [inline src/cascadia/PublicTerminalCore/HwndTerminal.cpp:44]: `CodepointWidthDetector` defaults to narrow glyphs if no fallback is specified already, so there's no need to do this.

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 5]
```diff
@@ -64,25 +64,6 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
-    {
-        // use C++11 magic statics to make sure we only do this once.
-        static auto initialized = []() {
-            // *** THIS IS A SINGLETON ***
-            SetGlyphWidthFallback(_IsGlyphWideForceNarrowFallback);
-
-            return true;
-        }();
-        return initialized;
-    }
```
Comment (lhecker) [inline src/cascadia/TerminalControl/ControlCore.cpp:84]: `CodepointWidthDetector` defaults to narrow glyphs if no fallback is specified already, so there's no need to do this.

File: src/host/ut_host/CodepointWidthDetectorTests.cpp [hunk 6]
```diff
@@ -38,18 +38,6 @@ class CodepointWidthDetectorTests
...
-    {
-        CodepointWidthDetector widthDetector;
-        for (const auto& data : testData)
-        {
-            const auto& expected = std::get<0>(data);
-            const auto& wstr = std::get<1>(data);
-            const auto result = widthDetector._extractCodepoint({ wstr.c_str(), wstr.size() });
-            VERIFY_ARE_EQUAL(result, expected);
-        }
-    }
```
Comment (lhecker) [inline src/host/ut_host/CodepointWidthDetectorTests.cpp:51]: This is already covered by the `CanGetWidths` test. `_extractCodepoint` doesn't exist anymore.

File: src/types/CodepointWidthDetector.cpp [hunk 7]
```diff
@@ -9,392 +9,358 @@ namespace
     // used to store range data in CodepointWidthDetector's internal map
     struct UnicodeRange final
     {
-        unsigned int lowerBound;
-        unsigned int upperBound;
-        CodepointWidth width;
+        char32_t lowerBound;
+        char32_t upperBound : 31;
+        char32_t isAmbiguous : 1;
```
Comment (lhecker) [inline src/types/CodepointWidthDetector.cpp:14]: This reduces the size of the table by 33%.

File: src/types/CodepointWidthDetector.cpp [hunk 8]
```diff
@@ -9,392 +9,358 @@ namespace
...
-        UnicodeRange{ 0x1fabf, 0x1fac5, CodepointWidth::Wide },
-        UnicodeRange{ 0x1face, 0x1fadb, CodepointWidth::Wide },
-        UnicodeRange{ 0x1fae0, 0x1fae8, CodepointWidth::Wide },
-        UnicodeRange{ 0x1faf0, 0x1faf8, CodepointWidth::Wide },
-        UnicodeRange{ 0x20000, 0x2fffd, CodepointWidth::Wide },
-        UnicodeRange{ 0x30000, 0x3fffd, CodepointWidth::Wide },
-        UnicodeRange{ 0xe0100, 0xe01ef, CodepointWidth::Ambiguous },
-        UnicodeRange{ 0xf0000, 0xffffd, CodepointWidth::Ambiguous },
-        UnicodeRange{ 0x100000, 0x10fffd, CodepointWidth::Ambiguous },
+    static constexpr std::array<UnicodeRange, 297> s_wideAndAmbiguousTable{
```
Comment (lhecker) [inline src/types/CodepointWidthDetector.cpp:27]: I've made it so that narrow codepoints aren't stored in this table anymore. This removes 3 entries.

File: src/types/CodepointWidthDetector.cpp [hunk 9]
```diff
@@ -9,392 +9,358 @@ namespace
...
     }
-    CATCH_LOG();
 
-    return true;
+    // The return value of _lookupGlyphWidth coincides with the enum value of CodepointWidth
+    // on purpose to allow for this easy conversion to happen. Optimally, we should probably
+    // remove CodepointWidth altogether to allow for ZWJs and other characters.
+    static_assert(WI_EnumValue(CodepointWidth::Narrow) == 1);
+    static_assert(WI_EnumValue(CodepointWidth::Wide) == 2);
+    return static_cast<CodepointWidth>(_lookupGlyphWidth(codepoint, glyph));
```
Comment (lhecker) [inline src/types/CodepointWidthDetector.cpp:363]: This code all the way to the end of the file is a complete rewrite. You can try reading it in the diff viewer, but there's unfortunately nothing to compare it with in the old code, outside of a few simple lines/comments. 😓

File: tools/Generate-CodepointWidthsFromUCD.ps1 [hunk 10]
```diff
@@ -224,20 +222,10 @@ $UCDRepertoire = $InputObject.ucd.repertoire.ChildNodes | Sort-Object {
     }
 }
 
-If (-not $Full) {
-    $UCDRepertoire = $UCDRepertoire | Where-Object {
-        # Select everything Wide/Ambiguous/Full OR Emoji w/ Emoji Presentation
-        ($_.ea -notin "N", "Na", "H") -or ($_.Emoji -eq "Y" -and $_.EPres -eq "Y")
-    }
-}
```
Comment (lhecker) [inline tools/Generate-CodepointWidthsFromUCD.ps1:232]: See `RemoveAll` below.

File: tools/Generate-CodepointWidthsFromUCD.ps1 [hunk 11]
```diff
@@ -260,21 +248,29 @@ If (-not $NoOverrides) {
     }
 }
 
+$ranges.RemoveAll({ $args[0].Width -eq [CodepointWidth]::Narrow }) | Out-Null
```
Comment (lhecker) [inline tools/Generate-CodepointWidthsFromUCD.ps1:251]: Moving this down, past the `If (-not $NoOverrides) {` block ensures that we filter out those 3 narrow codepoint overrides. Also I felt like this `RemoveAll` call would be more robust than the `Where-Object` block above.

File: tools/Generate-CodepointWidthsFromUCD.ps1 [hunk 12]
```diff
@@ -260,21 +248,29 @@ If (-not $NoOverrides) {
     }
 }
 
+$ranges.RemoveAll({ $args[0].Width -eq [CodepointWidth]::Narrow }) | Out-Null
+
+$c = 0
+ForEach($_ in $ranges) {
+    $c += $_.End - $_.Start + 1
+}
```
Comment (lhecker) [inline tools/Generate-CodepointWidthsFromUCD.ps1:256]: Counting the number of covered codepoints last ensures we pick up any changes due to `RemoveAll`.

File: src/types/GlyphWidth.cpp (hunk 1 above)
Comment (lhecker) [reply src/types/GlyphWidth.cpp:38]: Now that VS 17.3 includes a fix which makes passing 128-bit structs much faster on x64 in many situations, this is a huge "it depends" now. But before that, passing by reference was almost always faster. This is because the MSVC compiler often copied those parameters onto the stack with SSE instructions even if it didn't need to and 1-2 regular `mov`s would've been sufficient.

File: src/host/ut_host/CodepointWidthDetectorTests.cpp [hunk 13]
```diff
@@ -19,7 +19,7 @@ static const std::vector<std::tuple<unsigned int, std::wstring, CodepointWidth>>
     { 0x7, L"\a", CodepointWidth::Narrow }, // BEL
     { 0x20, L" ", CodepointWidth::Narrow },
     { 0x39, L"9", CodepointWidth::Narrow },
-    { 0x414, L"\x414", CodepointWidth::Ambiguous }, // U+0414 cyrillic capital de
+    { 0x414, L"\x414", CodepointWidth::Narrow }, // U+0414 cyrillic capital de
```
Comment (lhecker) [inline src/host/ut_host/CodepointWidthDetectorTests.cpp:22]: `widthDetector.GetWidth` won't ever return anything but `Narrow` or `Wide` from now on.

Comment (DHowett) [reply src/host/ut_host/CodepointWidthDetectorTests.cpp:22]: Then what we have is no longer a unit test! That's a functional test ;(

File: src/types/GlyphWidth.cpp (hunk 1 above)
Comment (KindDragon) [reply src/types/GlyphWidth.cpp:38]: Maybe my example is wrong, but looks like same instructions would generated by MSVC https://godbolt.org/z/W9P9WboEq

Comment (KindDragon) [reply src/types/GlyphWidth.cpp:38]: I joust found way to reproduce this old behaviour https://godbolt.org/z/vqqWjj8d8

Comment (lhecker) [reply src/types/GlyphWidth.cpp:38]: Yep. The issue used to occur if string-views are passed by-value by just a single level of indirection: https://godbolt.org/z/vh8c3sa17

This issue is especially glaring when used with the STL where all string-views are passed by-value externally and internally and you couldn't avoid this issue: https://godbolt.org/z/cPra48qqY
I'm not sure if our CI is on MSVC 19.33 already but if it is I think we can safely stop passing string-views by-reference. 🙂

Comment (KindDragon) [reply src/types/GlyphWidth.cpp:38]: Thanks for information

File: src/host/ut_host/CodepointWidthDetectorTests.cpp (hunk 13 above)
Comment (lhecker) [reply src/host/ut_host/CodepointWidthDetectorTests.cpp:22]: Eh, I'd say it's more like a conversion from a white-box to a black-box test. In this case it's fine though, since another test right after checks whether ambiguous characters result in the callback being called and being resolved correctly.

**Review by zadjii-msft — APPROVED** [review]

File: src/types/CodepointWidthDetector.cpp (hunk 7 above)
Comment (DHowett) [reply src/types/CodepointWidthDetector.cpp:14]: I love this.

**Review by DHowett — APPROVED** [review]
okay, fix the `-Full:` thing and we're good!

File: src/types/GlyphWidth.cpp [hunk 14]
```diff
@@ -22,7 +22,7 @@ bool IsGlyphFullWidth(const std::wstring_view glyph)
 //      wide or not. See CodepointWidthDetector::IsWide
 bool IsGlyphFullWidth(const wchar_t wch) noexcept
 {
-    return widthDetector.IsWide(wch);
+    return wch < 0x80 ? false : IsGlyphFullWidth({ &wch, 1 });
```
[the remark this answers is not in the document]
Comment (lhecker) [reply src/types/GlyphWidth.cpp:25]: Technically for performance it should help quite a bit, by allowing `IsGlyphFullWidth` to be inlined and the construction of the string-view and call to the other `IsGlyphFullWidth` to be skipped most of the time (just like most text is ASCII).
